source/poggles/texture.cpp

```diff
@@ -1,94 +1,125 @@
-               width,
-               height,
-               0,
-               format,
-               GL_UNSIGNED_BYTE,
-               data);
+poggles::texture::texture(GLenum target)
+    : m_original_target(target)
+{
+  bind();
 }
 
-void poggles::texture::load(GLenum target,
-                            float* data,
-                            int width,
-                            int height,
-                            int channels,
-                            GLint level)
+void poggles::texture::load(poggles::TextureType type)
 {
-  GLenum format = GL_RGB;
-  if (channels == 1) {
-    format = GL_RED;
-  } else if (channels == 2) {
-    format = GL_RG;
-  } else if (channels == 3) {
-    format = GL_RGB;
-  } else if (channels == 4) {
-    format = GL_RGBA;
-  }
-  glTexImage2D(target,
-               level,
-               static_cast<GLint>(format),
-               width,
-               height,
-               0,
-               format,
-               GL_FLOAT,
-               data);
-  // glGenerateMipmap(target);
+  std::visit(
+      overload {
+          [this](TextureFromFile const& textureFromFile)
+          {
+            int width = 0;
+            int height = 0;
+            int channels = 0;
+            uint8_t* data = stbi_load(textureFromFile.filename.c_str(),
+                                      &width,
+                                      &height,
+                                      &channels,
+                                      0);
+            if (data != nullptr) {
+              // Call ourselves, but with the relevant data extracted from the
+              // file
+              TextureFromData textureFromData {
+                  data,
+                  width,
+                  height,
+                  channels,
+                  textureFromFile.level,
+                  textureFromFile.target.value_or(m_original_target)};
+              load(textureFromData);
+            } else {
+              throw std::runtime_error("Failed to load texture");
+            }
+            stbi_image_free(data);
+          },
+          [this](TextureFromData const& textureFromData)
+          {
+            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT);
+            glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT);
+
+            GLenum format = GL_RGB;
+            if (textureFromData.channels == 1) {
+              format = GL_RED;
+            } else if (textureFromData.channels == 2) {
+              format = GL_RG;
+            } else if (textureFromData.channels == 3) {
+              format = GL_RGB;
+            } else if (textureFromData.channels == 4) {
+              format = GL_RGBA;
+            }
+            GLenum target = textureFromData.target.value_or(m_original_target);
```

Comment:
m_originalTarget doesn't get updated in the texture loading, and there's no check to make sure that m_originalTarget is the same as GL_TEXTURE_2D. This is a problem in the original code, and definitely not something you introduced, but it's something we should probably fix :grimacing: 

```suggestion       
            m_originalTarget = textureFromData.target.value_or(m_original_target);

            glTexParameteri(m_originalTarget, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
            glTexParameteri(m_originalTarget, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
            glTexParameteri(m_originalTarget, GL_TEXTURE_WRAP_S, GL_REPEAT);
            glTexParameteri(m_originalTarget, GL_TEXTURE_WRAP_T, GL_REPEAT);

            GLenum format = GL_RGB;
            if (textureFromData.channels == 1) {
              format = GL_RED;
            } else if (textureFromData.channels == 2) {
              format = GL_RG;
            } else if (textureFromData.channels == 3) {
              format = GL_RGB;
            } else if (textureFromData.channels == 4) {
              format = GL_RGBA;
```

Reply:
This is out out date now :)